### ISSUE TITLE
Envoy proxy memory usage spikes on service deployments

### ISSUE THREAD

**Issue by abhimoondra** [envoyproxy/envoy#30986]
*Description*:
Using envoy as a proxy for our services. The envoy proxy is running inside a Kubernetes pod along with a service which acts as its control plane. All our services also run in Kubernetes and this envoy proxy acts as an ingress.

We are facing memory usage spikes in the pods where envoy proxy was running. This was happening whenever a new pod of any service was getting added or deleted.

![iEnvoy Memory Usage](https://github.com/envoyproxy/envoy/assets/132544457/fb4064ec-ffcf-49c0-bede-a6f0924942ee)

On deeper analysis, I found out that we were updating our envoy config, for every pod addition/removal, which caused these memory spikes.

![image](https://github.com/envoyproxy/envoy/assets/132544457/c2b7f936-cc8d-42e5-b086-fbf672277bc8)

I did some research and found out that instead of updating the whole config we can just update the Delta. [Envoy Documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/xds_api#delta-endpoints)

After making these changes, here is what our config looks like (some part of it):

```
    "dynamic_resources": {
     "lds_config": {
      "api_config_source": {
       "api_type": "DELTA_GRPC",
       "grpc_services": [
        {
         "envoy_grpc": {
          "cluster_name": "metadata_service_xds"
         }
        }
       ],
       "transport_api_version": "V3"
      },
      "resource_api_version": "V3"
     },
     "cds_config": {
      "api_config_source": {
       "api_type": "DELTA_GRPC",
       "grpc_services": [
        {
         "envoy_grpc": {
          "cluster_name": "metadata_service_xds"
         }
        }
       ],
       "transport_api_version": "V3"
      },
      "resource_api_version": "V3"
     }
    },
    "admin": {
     "access_log_path": "/tmp/admin_access.log",
     "address": {
      "socket_address": {
       "address": "0.0.0.0",
       "port_value": 9901
      }
     }
    }
   }
```

Now in the envoy logs I can see that we are not updating all the clusters every time a new pod gets added, but I can see that the listeners still gets updated. Could anyone please help me if I am missing something?

![image](https://github.com/envoyproxy/envoy/assets/132544457/fa81e087-de92-41a1-8308-38d939fb7498)

**Comment by KBaichoo** [comment]
Does the memory still behave as above given the switch to delta? It also seems you have quite a bunch of grpc transcoder filters in the same filter chain -- how is it being used?

cc @htuch @adisuissa 

A heap profile https://github.com/envoyproxy/envoy/blob/main/bazel/PPROF.md#cpu-or-memory-consumption-testing-with-gperftools-and-pprof would be useful.

**Comment by abhimoondra** [comment]
Thank you for your response.

Yes, even after switching to delta the memory usage behavior remains same. 

Usecase of gRPC Json Transcorder: So we have a lot of services exposing gRPC endpoints. We have a internal tool which our development teams use to query different service endpoints to test etc., that tool doesn't support gRPC endpoints and hence we have these gRPC json transcoder filters for each service.

 Do you suggest splitting these gRPC trancoder filters in different filter chains and test?
 
 I will work on getting the heap profile meanwhile.

**Comment by adisuissa** [comment]
First I suggest considering using [ADS](https://www.envoyproxy.io/docs/envoy/v1.28.0/configuration/overview/xds_api#config-overview-ads) so the listeners and clusters are all passed on a single gRPC stream. Delta will probably still be a better solution.

IIRC listeners should only be updated if there's something different in the listener (i.e., it's hash is different). If listeners are constantly updated, then it may incur additional CPU and memory usage in Envoy (draining listeners).
If it's possible to decouple the listeners and their routes, then sending [RDS](https://www.envoyproxy.io/docs/envoy/v1.28.0/configuration/http/http_conn_man/rds#config-http-conn-man-rds) updates may be beneficial for your use-case.

**Comment by abhimoondra** [comment]
Here is the heap profile for one of the runs:

[heap_profile.tar.gz](https://github.com/envoyproxy/envoy/files/13466397/heap_profile.tar.gz)

Envoy Binary: [drive link](https://drive.google.com/file/d/10UutQhQedgdLY_MvwOyV2nQKZkf55I_j/view?usp=sharing)

**Stats:**

![image](https://github.com/envoyproxy/envoy/assets/132544457/bca7df80-5de8-4118-9da5-73cc354f2c40)


I ran the test for 15 mins, where I added few pods and removed few pods and envoy memory usage went from 800MB to 2.75 GB. There was no traffic served from the envoy proxy.

Our listener drain timeout is the default setting of 600 secs.In our production environment we generally have deployments every 2-5 mins.

One more thing that I wanted to point out was we were using an older version of envoy before but have upgraded to v1.28.0, even then we are seeing a similar behaviour.

@adisuissa I did made the changes to use ADS but that didn't have any effect on the memory usage. It is till the same.

**Comment by adisuissa** [comment]
> @adisuissa I did made the changes to use ADS but that didn't have any effect on the memory usage. It is till the same.

The use of ADS won't change much, other than optimize how many streams are created with the management server.
Using RDS (assuming you only need to update these routing table, rather than the listeners) will probably improve some of the memory usage.

**Comment by htuch** [comment]
I think two config dumps, one before and one after an update, might be useful.

**Comment by abhimoondra** [comment]
Sure here are the two config dumps. One when envoy memory usage was around 120MB and after scaling up and down few service pods, memory usage went up to 1.5 GB.

[envoy-120MB.json](https://github.com/envoyproxy/envoy/files/13503081/envoy-120MB.json)

[envoy-1.5GB.json](https://github.com/envoyproxy/envoy/files/13503085/envoy-1.5GB.json)

**Comment by htuch** [comment]
When I look at the diff between the JSON files (e.g. with `diff <(jq --sort-keys . envoy-120MB.json) <(jq --sort-keys . envoy-1.5GB.json)`), the bulk of the diff seems to be changes to the gRPC transcoder proto descriptors in binary form, i.e. `proto_descriptor_bin` field.

Is there a reason these are changing? 

I can imagine if we keep incrementing the size of the proto descriptor pool in memory over time, this could happen. I'm thinking of a scenario in which there are longer lived connections that are still using the old transcoder filter while the updates occur. This is consistent with a 600 second listener drain and your graphs spiking during rapid LDS update and then memory use dropping back down within the listener drain period when stabilized.

@nareddyt @lizan as CODEOWNERS.

**Comment by abhimoondra** [comment]
@htuch 

I think the command to check diff based on sorting the keys will not give the intended results as the gRPC JSON Transcoder filters are in an array and those won't be sorted.
We are not modifying the proto descriptors unless the underlying service changes its proto definition (which happens not  frequently).

Also, I was able to reproduce this when there was no traffic handled via envoy. I was just scaling up a few pods and memory usage was increasing on an isolated envoy proxy. So the long living connections are non-existent.

I do see logs in envoy about draining filter chains, could those be hanging on to these descriptors

**Comment by htuch** [comment]
Can you share the heap profile after pprof analysis, as a flamegraph or some kind of percentage breakdown?

**Comment by abhimoondra** [comment]
Sharing the screenshot of the flamegraph and some other views. I am unsure which one will be more helpful.

Flame Graph:
<img width="1728" alt="Flame Graph" src="https://github.com/envoyproxy/envoy/assets/132544457/6408edc4-8673-43c8-b689-4aae7bef2da5">

<img width="1716" alt="Screenshot 2023-12-09 at 3 28 12 PM" src="https://github.com/envoyproxy/envoy/assets/132544457/e930d237-4c76-4dc2-8624-43cf152b48e9">

Tree:
<img width="1716" alt="Part 1" src="https://github.com/envoyproxy/envoy/assets/132544457/338f1a88-43c8-4727-bdee-6e5d4d42438a">

<img width="1716" alt="Screenshot 2023-12-09 at 3 21 41 PM" src="https://github.com/envoyproxy/envoy/assets/132544457/d1b79b5e-c00c-4d9f-86b3-3fd22f814fc6">

Top:
<img width="1716" alt="Screenshot 2023-12-09 at 3 27 47 PM" src="https://github.com/envoyproxy/envoy/assets/132544457/1ec68ead-9f89-483a-862c-ac11827b4a6b">


Not sure which one will be most useful, so sharing whatever I could find. I have also attached envoy binary which I had used to generated the above heap profiles, just in case the above screenshots are not useful and you want to analyze heap profiles by youself.
https://github.com/envoyproxy/envoy/issues/30986#issuecomment-1826740152

**Comment by htuch** [comment]
Yeah, this confirms the hypothesis that memory is dominated by proto descriptors. I think if you reduce your drain time it should improve. CC @nareddyt @lizan for thoughts on how to improve this as CODEOWNERS, there's probably some smart descriptor sharing / reference counting / keying we could do.

**Comment by nareddyt** [comment]
@abhimoondra is each proto descriptor per gRPC-JSON transcoder filter completely different? Looking over [envoy-1.5GB.json](https://github.com/envoyproxy/envoy/files/13503085/envoy-1.5GB.json), the base64 contents do look different, but I'm curious if any large subset of protos are shared by multiple different descriptors.

If all the protos are very different, TBH this looks like intended behavior. Your config of protobuf descriptors serialized & encoded in base64 is 1.5 GB. I'm not surprised that parsing them out to our internal structures results in ~3 GB of memory usage. I'm assuming you're not seeing any memory leaks (where resource usage keeps growing even after stopping updates).

If you do have any underlying protos that are shared between all descriptors, I'd recommend you consolidate all gRPC transcoder config into one single filter. You can use protoc or protobuf toolchains to merge descriptors together. In theory, this should de-duplicate any shared protos in the descriptors, and make your config size much smaller.

**Comment by abhimoondra** [comment]
@nareddyt each proto descriptor per gRPC-JSON transcoder filter is different. Every service defines it's own proto contract and in our envoy config we store the base64 encoded version of that proto contract. So there are not much shared protos between the filters.

We are not seeing a memory leak rather memory usage spikes, where on deployments we see a spike in memory usage and after some time it subsides (basically when the filter chains have drained).

What I see is that even when there is no change in any of the proto descriptors' content for any of the gRPC-JSON transcoder filters, the memory usage increases. Is this correct behaviour?
Should the memory usage increase even when the proto descriptor content hasn't changed?

**Comment by abhimoondra** [comment]
@nareddyt could you please share any feedback on this? We are still seeing the issue.

**Comment by nareddyt** [comment]
> Should the memory usage increase even when the proto descriptor content hasn't changed?

IIUC this is occurring because the entire listener is being re-created with a new config. I think each individual transcoder config doesn't matter - if you change any section of the listener, the entire listener and all filters get re-created (?) Please have other Envoy exports double check me on this.

If that is true, this is expected behavior. During new listener config, a new listener is started with all new transcoder filters. Note the previous transcoder filters still exist for `drain-timeout-s`. During this time, we basically have duplicate filters, leading to 2x memory usage. After `drain-timeout-s` elapses, the older transcoder filters are cleaned up and memory usage goes back to normal.

This should be a general Envoy filter problem, not anything specific to the transcoder. It's just visible here because transcoder configs are huge.

Ref: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/draining

**Comment by abhimoondra** [comment]
Just an update, I implemented a fix on the similar lines of what is mentioned in [this](https://youtu.be/rdJZk6k314k?t=34) talk.

Before triggering an update to the Envoy configuration, I sorted the gRPC JSON transcoder filters according to the service name. This approach appears to have resolved the spikes in memory usage. Instead of observing the addition of new listeners and the draining of old ones with every deployment, I now notice these changes only when there is an actual modification in the proto descriptor (which I think is expected).

![image](https://github.com/envoyproxy/envoy/assets/132544457/8645dfe3-abb4-42db-bec4-dd9e8eb53d76)


Thank you for all your help. @htuch @nareddyt @adisuissa
